internal/api/oidc/key_test.go

```diff
@@ -0,0 +1,243 @@
+			alg:    "alg",
+			use:    domain.KeyUsageSigning,
+			seq:    1,
+			expiry: clock.Now().Add(time.Minute),
+		},
+		"key2": {
+			id:     "key2",
+			alg:    "alg",
+			use:    domain.KeyUsageSigning,
+			seq:    3,
+			expiry: clock.Now().Add(10 * time.Hour),
+		},
+	}
+)
+
+func queryKeyDB(_ context.Context, keyID string, current time.Time) (query.PublicKey, error) {
+	if key, ok := keyDB[keyID]; ok {
+		return key, nil
+	}
+	return nil, errors.New("not found")
+}
+
+func Test_keySetCache(t *testing.T) {
+	background, cancel := context.WithCancel(
+		clockwork.AddToContext(context.Background(), clock),
+	)
+	defer cancel()
+
+	// create an empty keySet with a purge go routine, runs every Hour
+	keySet := newKeySet(background, time.Hour, queryKeyDB)
+	ctx := authz.NewMockContext("instanceID", "orgID", "userID")
+
+	// query error
+	_, err := keySet.getKey(ctx, "key9")
+	require.Error(t, err)
+
+	want := &jose.JSONWebKey{
+		KeyID:     "key1",
+		Algorithm: "alg",
+		Use:       domain.KeyUsageSigning.String(),
+	}
+
+	// get key first time, populate the cache
+	got, err := keySet.getKey(ctx, "key1")
+	require.NoError(t, err)
+	assert.Equal(t, want, got)
+
+	// move time forward
+	clock.Advance(5 * time.Minute)
+	time.Sleep(time.Millisecond)
+
+	// key should still be in cache
+	keySet.mtx.RLock()
+	_, ok := keySet.instanceKeys["instanceID"]["key1"]
+	require.True(t, ok)
+	keySet.mtx.RUnlock()
+
+	// the key is expired, should error
+	_, err = keySet.getKey(ctx, "key1")
+	require.Error(t, err)
+
+	want = &jose.JSONWebKey{
+		KeyID:     "key2",
+		Algorithm: "alg",
+		Use:       domain.KeyUsageSigning.String(),
+	}
+
+	// get the second key from DB
+	got, err = keySet.getKey(ctx, "key2")
+	require.NoError(t, err)
+	assert.Equal(t, want, got)
+
+	// move time forward
+	clock.Advance(time.Hour)
+	time.Sleep(time.Millisecond)
+
+	// first key shoud be purged, second still present
+	keySet.mtx.RLock()
+	_, ok = keySet.instanceKeys["instanceID"]["key1"]
+	require.False(t, ok)
+	_, ok = keySet.instanceKeys["instanceID"]["key2"]
+	require.True(t, ok)
+	keySet.mtx.RUnlock()
+
+	// get the second key from cache
+	got, err = keySet.getKey(ctx, "key2")
+	require.NoError(t, err)
+	assert.Equal(t, want, got)
+
+	// move time forward
+	clock.Advance(10 * time.Hour)
+	time.Sleep(time.Millisecond)
+
+	// now the cache should be empty
+	keySet.mtx.RLock()
+	assert.Empty(t, keySet.instanceKeys)
+	keySet.mtx.RUnlock()
+}
+
+func Test_keySetCache_VerifySignature(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	k := newKeySet(ctx, time.Second, queryKeyDB)
+
+	tests := []struct {
+		name string
+		jws  *jose.JSONWebSignature
+	}{
+		{
+			name: "invalid token",
+			jws:  &jose.JSONWebSignature{},
+		},
+		{
+			name: "key not found",
+			jws: &jose.JSONWebSignature{
+				Signatures: []jose.Signature{{
+					Header: jose.Header{
+						KeyID: "xxx",
+					},
+				}},
+			},
+		},
+		{
+			name: "verify error",
```

Comment:
Successful test?

Reply:
As in the exisiting code I couldn't find unit tests that do actual JWS verification (please point out if they do exist), my reasoning was the following:

- `jws.Verify()` is the final call in the return in `VerifySignature()`.
- By only testing error cases coverage is 100% already
- `jws.Verify()` is imported and thus not spefically in our scope to test.
- Different codepaths on obtaining the public key (cached or DB) are unit tested
- There is an integration test covering introspection and therefore the success of `VerifySignature()`

If we want to unittest this function we would need to mock:

- Creation of a keypair
- Encrypting the public key into storage (yes we encrypt public keys!)
- Creation of an Access Token

I would argue that is overkill in this case, given the above reasoning.

Reply:
Fully understandable, wil resolve the comment.